File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.py
from collections import defaultdict
def solution(edges):
    answer = []
    shapes = {'start': 0, 'donut': 0, 'stick': 0, 'loop': 0}
    edge_cnt = {'in': 0, 'out': 0}
    graph = defaultdict(lambda: edge_cnt.copy())

    for start, end in edges:
        graph[start]['out'] += 1
        graph[end]['in'] += 1


    for node in graph:

        in_edge = graph[node]['in']
        out_edge = graph[node]['out']

        if in_edge == 0 and out_edge >= 2:
            shapes['start'] = node
        elif out_edge == 0 and in_edge >= 1:
            shapes['stick'] += 1
        elif in_edge >= 2 and out_edge >= 2:
            shapes['loop'] += 1

    shapes['donut'] = graph[shapes['start']]['out'] - shapes['stick'] - shapes['loop']


    for node in shapes:
        answer.append(shapes[node])

    return answer

File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.java
import java.util.*;
class Solution {
    public int[] solution(int[][] data){
        int n = data.length;
        // 0: 노드 번호, 1: 도넛, 2: 막대, 3: 8자
        int[] answer = new int[4];
        Map<Integer, int[]> graph = new HashMap<>(); // node, [in, out]
        for(int[] node : data){
            graph.computeIfAbsent(node[0], k->new int[2])[1]+=1;
            graph.computeIfAbsent(node[1], k->new int[2])[0]+=1;
        }

        for(int key : graph.keySet()){
            int[] val = graph.get(key);
            int in = val[0];
            int out = val[1];
            if(in == 0 && out >= 2) answer[0] = key;
            else if(out == 0 && in >= 1) answer[2]++;
            else if(out >= 2 && in >= 2) answer[3]++;
        }
        answer[1] = graph.get(answer[0])[1] - answer[2] - answer[3];
        System.out.println(Arrays.toString(answer));
        return answer;
    }
}